tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/OptimizerCore.scala

```diff
@@ -2183,6 +2216,43 @@ private[optimizer] abstract class OptimizerCore(
           case _               => default
         }
 
+      case String_+ =>
+        (lhs, rhs) match {
+          case (l: Literal, r: Literal) => 
```

Comment:
OK now I think I see how you can factor that much more. Instead of defining `literalToString(Literal): String`, define `foldToStringForString_+(tree: Tree): Tree` which transforms `IntLiteral(v)` into `StringLiteral(v.toString)`, for example, and returns `tree` if it cannot fold the conversion of `tree` to string.

Then, here, you can just do:
```scala
val lhs1 = foldToStringForString_+(lhs)
val rhs1 = foldToStringForString_+(rhs)
(lhs1, rhs1) match {
  ...
}
```
and now you know that you only have to deal with `StringLiteral`s. Any other stuff that you find is not foldable to a `String`, and therefore you need not worry about it. Then you will have *way* less cases here.

(note: from there you can't use `default` anymore, since at least you need to use `lhs1` and `rhs1`, not `lhs` and `rhs`, so `default`s should become `BinaryOp(String_+, lhs1, rhs1)` instead.